Feature/classic dimension

mvo5 opened the pull request:
This is the first part of the classic dimension work.

The branch adds a new classic package that has:
- classic.Create()
- classic.RunShell()
- classic.Destroy()

Until the rest of the work is done the classic mode is available via the following CLI:
 - sudo snappy enable-classic
 - sudo snappy shell classic
 - sudo snappy destroy-classic

In a followup branch I plan to make classic appear like a regular snap, i.e. it will be listed
via snappy list and "sudo snappy install classic" will install it, "sudo snappy purge classic" will remove it.

I'm not sure how to best test RunShell() as its really not doing much and it would probably just be "mockBindMount" "mockCallSystemd" and checking if those got called. Happy to do that if you think there is value in it.

niemeyer commented on classic/create.go:
```diff
@@ -0,0 +1,254 @@
+	}
+	release := lsb.Codename
+
+	needle := fmt.Sprintf("ubuntu;%s;%s;default", release, arch)
```
Should this have an extra separator at the end?

mvo5 replied:
Yes, good catch!

niemeyer commented on classic/create.go:
```diff
@@ -0,0 +1,254 @@
+	if err != nil {
+		return "", err
+	}
+	url := fmt.Sprintf("%s%s%s", lxdBaseURL, dlPath, "rootfs.tar.xz")
```
Should this use path.Join to prevent breaking if upstream changes whether the path ends in a slash or not?

mvo5 replied:
Indeed, thank you. Fixed.

niemeyer commented on classic/create.go:
```diff
@@ -0,0 +1,254 @@
+	defer resp.Body.Close()
+
+	if resp.StatusCode != 200 {
+		return "", fmt.Errorf("failed to download %s: %v", url, resp.StatusCode)
```
This line does not set err, which means the defer above does not remove the file. There's an easy fix for that, though: if we name the return values (`(path string, err error)`) Go will ensure the err result variable is set even in this case, before the deferred function is run, and thus the defer above works as-is.

mvo5 replied:
Nice, thanks for finding this issue! Fixed.

niemeyer commented on classic/create.go:
```diff
@@ -0,0 +1,254 @@
+		return "", err
+	}
+
+	client := &http.Client{}
```
It'd be good to have a timeout here. The default is none.

mvo5 replied:
Thanks! I was not aware of this, added a 30s default timeout but happy to tweak the value. Firefox uses a value of 300s here but that seems a bit excessive especially since the downloader in classic will have to download around ~80MB of data.

mvo5 replied:
Hm, I reverted my timeout change again. It seems like the timeout is for the entire download. So we need something very big as the download is very big (in the 80MB range). The http transport has a "dial" timeout of 30s it seems so that is covered. It seems like ideally we would use something like http://curl.haxx.se/libcurl/c/curl_easy_setopt.html#CURLOPTLOWSPEEDTIME here? But happy to take your advice.

niemeyer commented on classic/create.go:
```diff
@@ -0,0 +1,254 @@
+	return nil
+}
+
+var newProgress = func() progress.Meter {
```
var newProgress = progress.NewTextProgress

?

mvo5 replied:
Silly me, indeed :) However looking at this again I noticed that I should pass the progress.Meter into Create() instead so that e.g. webdm can pass its own progress. I did this tiny refactor now.

niemeyer commented on classic/create.go:
```diff
@@ -0,0 +1,254 @@
+	src := "/run/resolvconf/resolv.conf"
+	dst := filepath.Join(dirs.ClassicDir, "/run/resolvconf/")
+	if err := helpers.CopyFile(src, dst, helpers.CopyFlagPreserveAll); err != nil {
+		return fmt.Errorf("failed to copy %s to %s", src, dst)
```
Should we suffix with the returned error?  Also, shouldn't err be saying that message itself? Feels like it has the proper context for it.

mvo5 replied:
Yes, helpers.CopyFile() has the knowledge to give proper context. I looked at it and it seems like the error reporting could be improved but I will create a new branch for this.

niemeyer commented on classic/create.go:
```diff
@@ -0,0 +1,254 @@
+}
+
+// Create creates a new classic shell envirionment
+func Create() error {
```
Nice. This function makes it look so easy. :-)

niemeyer commented on classic/enabled.go:
```diff
@@ -0,0 +1,32 @@
+
+// Enabled returns true if the classic mode is already enabled
+func Enabled() bool {
+	return helpers.FileExists(filepath.Join(dirs.ClassicDir, "etc", "apt", "sources.list"))
```
Doesn't feel like this one line is worth its own file. :-)  Do you have further plans for the file?

mvo5 replied:
Good point, its a bit overkill. I moved it into helpers.go now.

niemeyer commented on classic/helpers.go:
```diff
@@ -0,0 +1,78 @@
+
+func mountpoint(path string) bool {
+	err := exec.Command("mountpoint", path).Run()
+	// man-page: zero if the directory is a mountpoint, non-zero if not
```
Or if anything else at all went wrong. Is there any way to be more deterministic about it? Command output? What to do with actual errors here?

mvo5 replied:
I reworked this code to make it more robust. Its a bit tricky because mountpoint.c from util-linux is not a great interface. I think my version is pretty close to "as good as it gets". To improve further we would have to rewrite mountpoint.c in go (which may actually be a good idea).

niemeyer commented on classic/helpers.go:
```diff
@@ -0,0 +1,78 @@
+
+func bindmount(src, dstPath, remountArg string) error {
+	dst := filepath.Join(dirs.ClassicDir, dstPath)
+	// already mounted
```
The reason we need a comment here is likely because the function name isn't more clear by itself. `isMounted` or something similar might make it redundant, maybe.

mvo5 replied:
Indeed, isMounted() is much nicer, I renamed it.

niemeyer commented on cmd/snappy/cmd_enable_classic.go:
```diff
@@ -0,0 +1,84 @@
+	"github.com/ubuntu-core/snappy/logger"
+)
+
+type cmdEnableClassic struct {
```
`struct{}` is the usual convention for emptiness.

mvo5 replied:
Thanks, I fixed that now.

niemeyer commented on cmd/snappy/cmd_enable_classic.go:
```diff
@@ -0,0 +1,84 @@
+var longEnableClassicHelp = i18n.G("Enable the ubuntu classic dimension.")
+
+func init() {
+	_, err := parser.AddCommand("enable-classic",
```
Regarding your comment about transforming this into snappy install classic, can we please talk about this before you push it forward?  I'm slightly concerned that this might become a leaky abstraction, so would appreciating understanding what depth of emulation you have in mind.

niemeyer commented:
A few superficial comments, but really like this overall. Thanks Michael! :+1:

mvo5 commented:
Thanks a lot for the excellent review! I think I addressed the issues and I'm happy to talk about the UI further. The UI in this pull request is good enough to test and explore the feature (and may even be good enough for the final version but I will leave that decision to the UI experts :)

zyga commented on classic/create.go:
```diff
@@ -0,0 +1,254 @@
+	return w.Name(), err
+}
+
+var policyRc = []byte(`
```
Not being familiar with the topic I would love to know what this is for.

zyga commented on classic/helpers.go:
```diff
@@ -0,0 +1,121 @@
+	}
+
+	// do the actual mount
+	cmd := exec.Command("mount", "--make-rprivate", "--rbind", "-o", "rbind", src, dst)
```
While I know _some_ of the more modern mount bits I'd love to have a comment that explains why we use rprivate.

mvo5 replied:
Added some docs here too

zyga commented on cmd/snappy/cmd_classic.go:
```diff
@@ -0,0 +1,78 @@
+		i18n.G("Destroy the ubuntu classic dimension."),
+		&cmdDestroyClassic{})
+	if err != nil {
+		logger.Panicf("Unable to enable-classic: %v", err)
```
copy paste

mvo5 replied:
Thanks! fixed.

zyga commented on cmd/snappy/cmd_classic.go:
```diff
@@ -0,0 +1,78 @@
+		i18n.G("Enable the ubuntu classic dimension."),
+		&cmdEnableClassic{})
+	if err != nil {
+		logger.Panicf("Unable to enable-classic: %v", err)
```
Is this something that we want to do across the API down the line? (snap vs snappy)

zyga commented on cmd/snappy/cmd_classic.go:
```diff
@@ -0,0 +1,78 @@
+		return err
+	}
+
+	fmt.Println(`Classic dimension enabled on this snappy system.
```
i18n?

mvo5 replied:
Nice catch, thanks!

zyga commented on cmd/snappy/cmd_classic.go:
```diff
@@ -0,0 +1,78 @@
+		return err
+	}
+
+	fmt.Println(`Classic dimension destroyed on this snappy system.`)
```
i18n

mvo5 replied:
Fixed here too now.

zyga commented on cmd/snappy/cmd_shell.go:
```diff
@@ -0,0 +1,60 @@
+
+func init() {
+	arg, err := parser.AddCommand("shell",
+		"Run snappy shell interface",
```
i18n

mvo5 replied:
Thanks, fixed.

zyga commented on cmd/snappy/cmd_shell.go:
```diff
@@ -0,0 +1,60 @@
+		"Run snappy shell interface",
+		&cmdShell{})
+	if err != nil {
+		logger.Panicf("Unable to : %q", err)
```
"Unable to add shell command: %q"

mvo5 replied:
Thanks, fixed!

zyga commented on cmd/snappy/cmd_shell.go:
```diff
@@ -0,0 +1,60 @@
+		return classic.RunShell()
+	}
+
+	return fmt.Errorf("unsupported shell %v", shellType)
```
i18n

mvo5 replied:
Good catch!

zyga commented on classic/create.go:
```diff
@@ -0,0 +1,254 @@
+	}
+
+	// enable libnss-extrausers
+	if err := runInChroot(dirs.ClassicDir, "apt-get", "install", "-y", "libnss-extrausers"); err != nil {
```
Do we have any chance of blocking debconf prompts here?

mvo5 replied:
We do, we can set DEBIAN_FRONTEND=noninteractive in the environment. But I think thats not needed because the libnss-extrausers does not use debconf. If it becomes a problem its trivial to add :)

zyga commented on classic/create.go:
```diff
@@ -0,0 +1,254 @@
+	if err := runInChroot(dirs.ClassicDir, "apt-get", "install", "-y", "libnss-extrausers"); err != nil {
+		return err
+	}
+	cmd := exec.Command("sed", "-i", "-r", "/^(passwd|group|shadow):/ s/$/ extrausers/", filepath.Join(dirs.ClassicDir, "/etc/nsswitch.conf"))
```
Uh, a comment would be nice here

mvo5 replied:
Thanks, adding one

zyga commented on cmd/snappy/cmd_classic.go:
```diff
@@ -0,0 +1,78 @@
+	}
+
+	pbar := progress.NewTextProgress()
+	if err := classic.Create(pbar); err != nil {
```
Enabled but Create (not Enable)

mvo5 replied:
Hmmmmm, not sure, Created() instead of Enabled() feels a bit odd too. /me ponders

zyga replied:
I meant that you can have if !classic.Enabled() { classic.Enable() }

zyga commented:
overall +1 though, no need to block on anything I've mentioned. There's at least one small copy-paste bug but you can address that later too.

chipaca commented:
Looks good. Silly question: can't this be made three subcommands of a 'classic' command?

mvo5 commented:
@chipaca thanks! the UI question is still open I'm pretty sure this will not be the final UI. But we should have all the building blocks with this branch, see e.g. https://github.com/mvo5/snappy/tree/feature/classic-dimension-ui for one possible UI

chipaca commented on classic/create.go:
```diff
@@ -0,0 +1,260 @@
+// The format is specified in
+// https://people.debian.org/~hmh/invokerc.d-policyrc.d-specification.txt
+var policyRc = []byte(`
+#!/bin/sh
```
Doesn't this create a file which first character is \n? Dies that work?

mvo5 replied:
Good catch \o/ It does work, funny enough this works. Fixed anyway.

chipaca commented on classic/create.go:
```diff
@@ -0,0 +1,262 @@
+var policyRc = []byte(`
+#!/bin/sh
+while true; do
+    case "\$1" in
```
I'm pretty sure that \ is spurious

mvo5 replied:
And you are absolutely correct, thanks again! Fixed.

chipaca commented on classic/create.go:
```diff
@@ -0,0 +1,262 @@
+	cmdArgs = append(cmdArgs, cmd...)
+
+	if output, err := exec.Command(cmdArgs[0], cmdArgs[1:]...).CombinedOutput(); err != nil {
+		return fmt.Errorf("failed to run %s in chroot %s: %s (%s)", cmd, chroot, err, output)
```
You probably want ℅q instead of℅s

mvo5 replied:
Yes, done.

chipaca commented on classic/common_test.go:
```diff
@@ -0,0 +1,29 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2014-2015 Canonical Ltd
```
2015

mvo5 replied:
Indeed, thank you

chipaca commented on classic/create.go:
```diff
@@ -0,0 +1,262 @@
+
+	"github.com/ubuntu-core/snappy/arch"
+	"github.com/ubuntu-core/snappy/dirs"
+
```
What's this space for?

mvo5 replied:
No good reason, removed.

chipaca commented on classic/create.go:
```diff
@@ -0,0 +1,262 @@
+		return "", err
+	}
+
+	return "", fmt.Errorf("needle %s not found", needle)
```
%q

mvo5 replied:
+1

chipaca commented on classic/create.go:
```diff
@@ -0,0 +1,262 @@
+	}
+
+	if err := scanner.Err(); err != nil {
+		return "", err
```
A bit of context here would be good

mvo5 replied:
Thanks, added!

chipaca commented on classic/create.go:
```diff
@@ -0,0 +1,262 @@
+func findDownloadURL() (string, error) {
+	resp, err := http.Get(lxdBaseURL + lxdIndexPath)
+	if err != nil {
+		return "", fmt.Errorf("failed to downlaod lxdIndexUrl: %s", err)
```
%v

mvo5 replied:
Hm, I really need to make this a habit. Thanks and fixed.

zyga commented:
Thanks for addressing the things I've pointed out. I cannot wait to use this :)
